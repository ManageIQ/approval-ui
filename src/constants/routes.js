const routes = {
  allrequests: {
    index: '/allrequests'
  },
  requests: {
    index: '/requests',
    addComment: '/requests/add-comment',
    approve: '/requests/approve',
    deny: '/requests/deny'
  },
  request: {
    index: '/request',
    addComment: '/request/add-comment',
    approve: '/request/approve',
    deny: '/request/deny'
  },
  workflows: {
    index: '/workflows',
    add: '/workflows/add-workflow',
    remove: '/workflows/remove',
    editInfo: '/workflows/edit-info',
    editGroups: '/workflows/edit-groups',
    editSequence: '/workflows/edit-sequence'
  }
};

export default routes;
