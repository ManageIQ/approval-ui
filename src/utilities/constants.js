import {
  CheckCircleIcon,
  InfoCircleIcon,
  ErrorCircleOIcon,
  ExclamationCircleIcon
} from '@patternfly/react-icons';
import React from 'react';
import requestsMessages from '../messages/requests.messages';

export const APPROVAL_API_BASE = `${process.env.BASE_PATH}/approval/v1.2`;
export const RBAC_API_BASE = `${process.env.BASE_PATH}/rbac/v1`;

export const decisionValues = {
  undecided: { displayName: requestsMessages.needsReview, color: 'blue', icon: <InfoCircleIcon /> },
  approved: { displayName: requestsMessages.approved, color: 'green', icon: <CheckCircleIcon /> },
  denied: { displayName: requestsMessages.denied, color: 'red', icon: <ExclamationCircleIcon /> },
  canceled: { displayName: requestsMessages.canceled, color: 'red', icon: <ErrorCircleOIcon /> },
  error: { displayName: requestsMessages.error, color: 'red', icon: <ExclamationCircleIcon /> }
};

// React intl does not support empty strings
export const untranslatedMessage = (defaultMessage = ' ') => ({ id: 'untranslated', defaultMessage });
export const APP_DISPLAY_NAME = {
  catalog: 'Automation Services Catalog',
  topology: 'Topological inventory'
};
