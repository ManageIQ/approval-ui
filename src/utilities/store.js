
import promiseMiddleware from 'redux-promise-middleware';
import ReducerRegistry, { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/files/ReducerRegistry';
import { notifications, notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';

import reduxLogger from 'redux-logger';
import thunk from 'redux-thunk';
import requestReducer, { requestsInitialState } from '../redux/reducers/request-reducer';
import workflowReducer, { workflowsInitialState } from '../redux/reducers/workflow-reducer';
import groupReducer, { groupsInitialState } from '../redux/reducers/group-reducer';
import rolesReducer, { rolesInitialState } from '../redux/reducers/roles-reducer';

const registry = new ReducerRegistry({}, [ thunk, promiseMiddleware(), notificationsMiddleware({
  errorTitleKey: [ 'message' ],
  errorDescriptionKey: [ 'errors', 'stack' ]
}), reduxLogger ]);

registry.register({
  requestReducer: applyReducerHash(requestReducer, requestsInitialState),
  workflowReducer: applyReducerHash(workflowReducer, workflowsInitialState),
  groupReducer: applyReducerHash(groupReducer, groupsInitialState),
  rolesReducer: applyReducerHash(rolesReducer, rolesInitialState),
  notifications
});

export default registry.getStore();
