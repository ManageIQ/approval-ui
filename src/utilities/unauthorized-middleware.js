import { approvalHistory } from '../router';

const unAuthorizedMiddleware = () => (dispatch) => (action) => {
  const nextAction = { ...action };
  if (action.type.match(/_REJECTED$/) && action?.payload?.redirect) {
    setTimeout(() => {
      approvalHistory.push(action.payload.redirect.pathname, {
        from: approvalHistory.location
      });
    });
    return;
  }

  return dispatch(nextAction);
};

export default unAuthorizedMiddleware;
