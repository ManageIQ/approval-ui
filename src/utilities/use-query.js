import { useLocation } from 'react-router-dom';

const useQuery = (requiredParams = []) => {
  const { search } = useLocation();
  const query = new URLSearchParams(search);
  return [
    requiredParams.reduce(
      (acc, curr) => ({
        ...acc,
        [curr]: query.get(curr)
      }),
      {}
    ),
    search,
    query
  ];
};

export default useQuery;
