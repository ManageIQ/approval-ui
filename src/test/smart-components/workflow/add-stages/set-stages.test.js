import React from 'react';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';
import thunk from 'redux-thunk';
import SetStages from '../../../../smart-components/workflow/add-stages/set-stages';
import { TrashIcon } from '@patternfly/react-icons';
import AsyncSelect from 'react-select/async';
import { RBAC_API_BASE } from '../../../../utilities/constants';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store' ;
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications';

describe('<SetStages />', () => {
  let initialProps;
  const middlewares = [ thunk, promiseMiddleware(), notificationsMiddleware() ];
  let mockStore;

  const ComponentWrapper = ({ store, initialEntries, children }) => (
    <Provider store={ store }>
      <MemoryRouter initialEntries={ initialEntries } keyLength={ 0 }>
        { children }
      </MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    initialProps = {
      formData: {
        wfGroups: []
      },
      handleChange: jest.fn(),
      defaultOptions: [],
      title: 'Set stages test'
    };
    mockStore = configureStore(middlewares);
  });

  it('should call remove stage callback', async () => {
    const store = mockStore({ groupReducer: { groups: [{  value: '123', label: 'Group 1' }]}});
    let wrapper;
    const handleChange = jest.fn();
    apiClientMock.get(`${RBAC_API_BASE}/groups/`, mockOnce({ body: { data: []}}));
    await act(async() => {
      wrapper = mount(
        <ComponentWrapper store={ store } >
          <SetStages
            { ...initialProps }
            formData={ { wfGroups: [{ id: 'should be in callback' }, { id: 'should not be in callback' }]} }
            handleChange={ handleChange }
          />
        </ComponentWrapper>);
    });
    wrapper.update();

    wrapper.find('button.pf-c-button.pf-m-link').first().simulate('click');
    expect(handleChange).toHaveBeenCalledWith({ wfGroups: [{ id: 'should be in callback' }]});
  });

  it('should call add stage callback', async () => {
    const store = mockStore({ groupReducer: { groups: [{  value: '123', label: 'Group 1' }]}});
    apiClientMock.get(`${RBAC_API_BASE}/groups/`, mockOnce({ body: { data: []}}));
    let wrapper;
    await act(async() => {
      wrapper = mount(
        <ComponentWrapper store={ store } >
          <SetStages
            { ...initialProps }
            formData={ { wfGroups: []} }
          /></ComponentWrapper>);
    });
    wrapper.update();

    expect(wrapper.find(TrashIcon)).toHaveLength(0);
    /**
     * Need to add two stages, first can not bet deleted
     */
    wrapper.find('button.pf-c-button.pf-m-link').first().simulate('click');
    wrapper.find('button.pf-c-button.pf-m-link').first().simulate('click');
    wrapper.update();
    expect(wrapper.find(TrashIcon)).toHaveLength(1);
  });

  it('should call onInputChange callback', async (done) => {
    const store = mockStore({ groupReducer: { groups: [{  value: '123', label: 'Group 1' }]}});
    let wrapper;
    apiClientMock.get(`${RBAC_API_BASE}/groups/`, mockOnce({ body: { data: []}}));
    await act(async() => {
      wrapper = mount(
        <ComponentWrapper store={ store } >
          <SetStages
            { ...initialProps }
            formData={ { wfGroups: [{}]} }
          /></ComponentWrapper>);
    });
    wrapper.update();
    const selectInput = wrapper.find('input');
    selectInput.getDOMNode().value = 'foo';
    selectInput.simulate('change');
    wrapper.update();
    expect(wrapper.find(AsyncSelect).props().inpuValue).toEqual('foo');
    done();
  });

  it('should call loadOptions request for async select', async (done) => {
    expect.assertions(1);
    const store = mockStore({ groupReducer: { groups: [{  value: '123', label: 'Group 1' }]}});

    apiClientMock.get(`${RBAC_API_BASE}/groups/`, mockOnce({ body: { data: []}}));
    apiClientMock.get(`${RBAC_API_BASE}/groups/?name=foo`, mockOnce((req, res) => {
      expect(req).toBeTruthy();
      done();
      return res.status(200);
    }));
    let wrapper;
    await act(async() => {
      wrapper = mount(
        <ComponentWrapper store={ store } >
          <SetStages
            { ...initialProps }
            formData={ { wfGroups: [{}]} }
          /></ComponentWrapper>);
    });
    wrapper.update();

    const selectInput = wrapper.find('input');
    selectInput.getDOMNode().value = 'foo';
    selectInput.simulate('change');
    wrapper.update();
    setTimeout(() => {
      wrapper.update();
    }, 251);
  });

  it('should call handleChange prop', async (done) => {
    const store = mockStore({ groupReducer: { groups: [{  value: '123', label: 'Group 1' }]}});

    let wrapper;
    const handleChange = jest.fn();
    apiClientMock.get(`${RBAC_API_BASE}/groups/`, mockOnce({ body: { data: []}}));
    await act(async() => {
      wrapper = mount(
        <ComponentWrapper store={ store } >
          <SetStages
            { ...initialProps }
            formData={ { wfGroups: [{}]} }
            options={ [{ label: 'foo', value: '1' }] }
            handleChange={ handleChange }
          /></ComponentWrapper>);
    });
    wrapper.update();

    wrapper.find(AsyncSelect).props().onChange({ value: '1', label: 'foo' });
    expect(handleChange).toHaveBeenCalledWith({ wfGroups: [{ label: 'foo', value: '1' }]});
    done();
  });
});
