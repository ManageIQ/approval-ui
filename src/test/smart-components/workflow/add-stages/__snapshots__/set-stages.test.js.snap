// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SetStages /> should render correctly 1`] = `
<SetStages
  formData={
    Object {
      "wfGroups": Array [],
    }
  }
  handleChange={[MockFunction]}
  options={Array []}
  title="Set stages test"
>
  <Stack
    gutter="md"
  >
    <div
      className="pf-l-stack pf-m-gutter"
    >
      <StackItem>
        <div
          className="pf-l-stack__item"
        >
          <Title
            size="md"
          >
            <h1
              className="pf-c-title pf-m-md"
            >
              Set stages test
            </h1>
          </Title>
        </div>
      </StackItem>
      <StackItem>
        <div
          className="pf-l-stack__item"
        >
          <Stack
            gutter="sm"
          >
            <div
              className="pf-l-stack pf-m-gutter"
            >
              <StackItem
                style={
                  Object {
                    "borderTop": 10,
                  }
                }
              >
                <div
                  className="pf-l-stack__item"
                  style={
                    Object {
                      "borderTop": 10,
                    }
                  }
                >
                  <Component
                    id="add-workflow-stage"
                    isInline={true}
                    onClick={[Function]}
                    variant="link"
                  >
                    <ComponentWithOuia
                      component={[Function]}
                      componentProps={
                        Object {
                          "children": Array [
                            <PlusIcon
                              color="currentColor"
                              noVerticalAlign={false}
                              size="sm"
                              title={null}
                            />,
                            " ",
                            "Add a group",
                          ],
                          "id": "add-workflow-stage",
                          "isInline": true,
                          "onClick": [Function],
                          "variant": "link",
                        }
                      }
                      consumerContext={null}
                    >
                      <Button
                        id="add-workflow-stage"
                        isInline={true}
                        onClick={[Function]}
                        ouiaContext={
                          Object {
                            "isOuia": false,
                            "ouiaId": null,
                          }
                        }
                        variant="link"
                      >
                        <button
                          aria-disabled={null}
                          aria-label={null}
                          className="pf-c-button pf-m-link pf-m-inline"
                          disabled={false}
                          id="add-workflow-stage"
                          onClick={[Function]}
                          tabIndex={null}
                          type="button"
                        >
                          <PlusIcon
                            color="currentColor"
                            noVerticalAlign={false}
                            size="sm"
                            title={null}
                          >
                            <svg
                              aria-hidden={true}
                              aria-labelledby={null}
                              fill="currentColor"
                              height="1em"
                              role="img"
                              style={
                                Object {
                                  "verticalAlign": "-0.125em",
                                }
                              }
                              viewBox="0 0 448 512"
                              width="1em"
                            >
                              <path
                                d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                                transform=""
                              />
                            </svg>
                          </PlusIcon>
                           
                          Add a group
                        </button>
                      </Button>
                    </ComponentWithOuia>
                  </Component>
                </div>
              </StackItem>
            </div>
          </Stack>
        </div>
      </StackItem>
    </div>
  </Stack>
</SetStages>
`;
